File: bank-ocr/typescript/tests/AccountNumberBuilder.ts
import { AccountLength, DigitHeight } from '../src/constants.js';
import { DigitBuilder } from './DigitBuilder.js';

export class AccountNumberBuilder {
  private digits: string[][] = Array.from({ length: AccountLength }, () =>
    new DigitBuilder().forDigit(0).build(),
  );

  fromString(digits: string): this {
    if (digits.length !== AccountLength) {
      throw new Error(`Expected ${AccountLength} digits, got ${digits.length}.`);
    }
    this.digits = Array.from(digits).map((ch) =>
      new DigitBuilder().forDigit(Number.parseInt(ch, 10)).build(),
    );
    return this;
  }

  withDigitAt(position: number, glyph: string[]): this {
    this.digits[position] = [...glyph];
    return this;
  }

  buildRows(): string[] {
    const rows: string[] = [];
    for (let r = 0; r < DigitHeight; r++) {
      rows.push(this.digits.map((d) => d[r]!).join(''));
    }
    return rows;
  }
}

File: bank-ocr/python/tests/account_number_builder.py
from __future__ import annotations

from bank_ocr import AccountLength, DigitHeight

from .digit_builder import DigitBuilder


class AccountNumberBuilder:
    def __init__(self) -> None:
        self._digits: list[list[str]] = [
            DigitBuilder().for_digit(0).build() for _ in range(AccountLength)
        ]

    def from_string(self, digits: str) -> "AccountNumberBuilder":
        if len(digits) != AccountLength:
            raise ValueError(
                f"Expected {AccountLength} digits, got {len(digits)}."
            )
        self._digits = [
            DigitBuilder().for_digit(int(ch)).build() for ch in digits
        ]
        return self

    def with_digit_at(self, position: int, glyph: list[str]) -> "AccountNumberBuilder":
        self._digits[position] = list(glyph)
        return self

    def build_rows(self) -> list[str]:
        rows: list[str] = []
        for r in range(DigitHeight):
            rows.append("".join(d[r] for d in self._digits))
        return rows
